baseplate/lib/config.py

```diff
@@ -390,23 +390,40 @@ def tuple_of(text: str) -> Sequence[T]:
 
 
 def DefaultFromEnv(
-    item_parser: Callable[[str], T], default_src: str, fallback: OptionalType[T] = None
+    item_parser: Callable[[str], T], default_src: str, fallback: OptionalType[T] = None,
 ) -> Callable[[str], OptionalType[T]]:  # noqa: D401
     """An option of type T or a default.
 
     The default is sourced from an environment variable with the name specified in ``default_src``.
     If the environment variable is not set, then the fallback will be used.
     One of the following values must be provided: fallback, default_src, or the provided configuration
     """
-    env = os.getenv(default_src) or ""
-    default = Optional(item_parser, fallback)(env)
+    parser = OptionalDefaultFromEnv(item_parser, default_src, fallback)
 
     def default_from_env(text: str) -> OptionalType[T]:
-        val = Optional(item_parser, default)(text)
+        val = parser(text)
         if val:
             return val
 
-        raise ValueError("no value provided")
+        raise ValueError("Value is required.")
+
+    return default_from_env
+
+
+def OptionalDefaultFromEnv(
```

Comment:
Do we need this? I think we can just do it with composition: `Optional(DefaultFromEnv(...))`?

Reply:
The issue with using `Optional(DefaultFromEnv(...))` is that `DefaultFromEnv(...)` injects defaults. The problematic code in `Optional` is in the parser function:
```
if text:
    return item_parse(text)

return default
```

In the case of `DefaultFromEnv` we need the item parser to always parse text, not only when text is provided. We could re-write `Optional(...) like this to account for this:

```
try:
    return item_parser(text)
except ValueError:
    return default
```
There are a few downsides to this implementation. First, `ValueError` is not exclusively used to mean that a parser didn't receive a valid value, it's also used to indicate the parser received an invalid value. Also there is some odd semantics is using `Optional` with `DefaultFromEnv`, as `DefaultFromEnv` already has a fallback. This creates two layers of defaults.

What do you think is the best way forward here?

Reply:
Got it. Good point. What you have is probably best then.